bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/scheduling/Strand.java

```diff
@@ -256,14 +261,22 @@ public void handleWaitMultiple(Map<String, FutureValue> keyValues, MapValue targ
                     ctx.unLock();
                     throw future.panic;
                 }
+                if (future.hasWaited()) {
+                    target.put(StringUtils.fromString(entry.getKey()), createWaitOnSameFutureError());
+                } else {
+                    target.put(StringUtils.fromString(entry.getKey()), future.result);
+                }
                 ctx.waitCount.decrementAndGet();
-                target.put(StringUtils.fromString(entry.getKey()), future.result);
+                doneCount++;
             } else {
                 this.setState(BLOCK_ON_AND_YIELD);
                 entry.getValue().strand.waitingContexts.add(ctx);
             }
             future.strand.unlock();
         }
+        if (doneCount == keyValues.size()) {
+            keyValues.forEach((s, futureValue) -> futureValue.setWaited(true));
```

Comment:
let's use a simple loop, we have to convention of using raw loops for performance/debug critical paths.

Reply:
why is the setWaited happening  after unlocking, want this cause a issue?

Reply:
Created an issue to handle potential concurrency issues: https://github.com/ballerina-platform/ballerina-lang/issues/30621